File: joss/comparisons/cat-4-join-safety/17-join-nullability/example-join-nullability.R
# Error Class 17: Join Nullability
#
# After a left_join, unmatched rows get NA for right-side columns.
# R silently lets you operate on these NA values — arithmetic
# propagates NA, string methods produce NA, and comparisons return
# NA (which filter() silently drops).

library(tidyverse)

patients <- tibble(
  patient_id = c("P001", "P002", "P003"),
  name = c("Alice Johnson", "Bob Smith", "Carol Davis")
)

encounters <- tibble(
  patient_id = c("P001", "P001"),
  department = c("Emergency", "ICU"),
  los_days = c(3, 7)
)

# Left join — P002 and P003 have NA for department and los_days
joined <- patients %>% left_join(encounters, by = "patient_id")

# SILENT: toupper() on NA produces NA — no error, no warning
joined <- joined %>% mutate(dept_upper = toupper(department))
print(joined %>% select(patient_id, department, dept_upper))
# P002: NA → NA, P003: NA → NA

# SILENT: Arithmetic on NA propagates — no error, no warning
joined <- joined %>% mutate(los_weeks = los_days / 7)
print(joined %>% select(patient_id, los_days, los_weeks))
# P002: NA → NA, P003: NA → NA

# SILENT: filter() drops NA rows — no error, no warning
long_stays <- joined %>% filter(los_days > 5)
cat("Rows with los > 5:", nrow(long_stays), "\n")
# NA rows silently dropped

File: joss/comparisons/cat-4-join-safety/17-join-nullability/example-join-nullability.py
"""
Error Class 17: Join Nullability

After a left merge, unmatched rows get NaN for right-side columns.
Python silently lets you operate on these NaN values — arithmetic
propagates NaN, string methods produce NaN, and comparisons exclude
NaN rows without any warning.
"""
import pandas as pd

patients = pd.DataFrame({
    "patient_id": ["P001", "P002", "P003"],
    "name": ["Alice Johnson", "Bob Smith", "Carol Davis"],
})

encounters = pd.DataFrame({
    "patient_id": ["P001", "P001"],
    "department": ["Emergency", "ICU"],
    "los_days": [3, 7],
})

# Left merge — P002 and P003 have NaN for department and los_days
joined = patients.merge(encounters, on="patient_id", how="left")

# SILENT: .str.upper() on NaN produces NaN — no error, no warning
joined["dept_upper"] = joined["department"].str.upper()
print(joined[["patient_id", "department", "dept_upper"]])
# P002: NaN → NaN, P003: NaN → NaN

# SILENT: Arithmetic on NaN propagates — no error, no warning
joined["los_weeks"] = joined["los_days"] / 7
print(joined[["patient_id", "los_days", "los_weeks"]])
# P002: NaN → NaN, P003: NaN → NaN

# SILENT: Comparison excludes NaN rows — no error, no warning
long_stays = joined[joined["los_days"] > 5]
print(f"Rows with los > 5: {len(long_stays)}")
# NaN rows silently excluded from result
